cpp/src/arrow/compute/function.cc

```diff
@@ -426,6 +426,7 @@ Status ScalarFunction::AddKernel(ScalarKernel kernel) {
   if (arity_.is_varargs && !kernel.signature->is_varargs()) {
     return Status::Invalid("Function accepts varargs but kernel signature does not");
   }
+  kernel.is_pure = is_pure();
```

Comment:
The is_pure property need to be used in kernel's execute. I'm not sure should we add kernel.is_pure to scalarkernel's constructor, because there is a second way to emplace_back kernels_ in another AddKernel?

But actually the current way is used in only built-in impure functions (random) :).

Reply:
I've often thought about adding `const Function *Kernel::function` as a non owning reference from the kernel to its function, which would be a larger change but perhaps less of a special case than duplicating this one property. As a side benefit, it'd allow us to remove the [bound call expression property `function`](https://github.com/bkietz/arrow/blob/46758bc3c6321d8b7013acf52bff7761a8b33eda/cpp/src/arrow/compute/expression.h#L55) @lidavidm @pitrou what do you think?

Reply:
Would weak_ptr be more explicit? Otherwise makes sense

Reply:
Seems reasonable, so should we promote this refactoring in other PRs?

Reply:
> Would weak_ptr be more explicit?

I don't think weak_ptr adds anything; the lifetime of a Function is managed by the registry anyway, IE they are basically never destroyed. If ownership becomes a concern later it's easy enough to add `enable_shared_from_this` to `compute::Function` or replace with a weak_ptr

> Seems reasonable, so should we promote this refactoring in other PRs?

It does seem reasonable to open a separate PR to add `Kernel::function`

Reply:
> It does seem reasonable to open a separate PR to add `Kernel::function`

I've added an issue to track this https://github.com/apache/arrow/issues/40840